Sources/Db/APIs/MySQL.php

```diff
@@ -2020,23 +2020,26 @@ protected function __construct(array $options = [])
 			$this->prefixReservedTables();
 		}
 
-		$this->get_version();
-		$this->supports_pcre = version_compare($this->version, strpos($this->version, 'MariaDB') !== false ? '10.0.5' : '8.0.4', '>=');
+		if (empty($this->connection) && !$non_fatal) {
+			$this->get_version();
+			$this->supports_pcre = version_compare($this->version, strpos($this->version, 'MariaDB') !== false ? '10.0.5' : '8.0.4', '>=');
+		}
 
 		// For backward compatibility.
 		if (!is_object(self::$db_connection)) {
 			self::$db_connection = $this->connection;
 		}
 
 		// Ensure database has UTF-8 as its default input charset.
-		$this->query(
-			'',
-			'
-			SET NAMES {string:db_character_set}',
-			[
-				'db_character_set' => $this->character_set,
-			],
-		);
+		if (empty($this->connection) && !$non_fatal)
+			$this->query(
+				'',
+				'
+				SET NAMES {string:db_character_set}',
+				[
+					'db_character_set' => $this->character_set,
+				],
+			);
```

Comment:
As above, why have these lines been wrapped in this `if` block? The test here, `(empty($this->connection) && !$non_fatal)` will never be true, because if it were we would already have bailed out higher up. As a result, the lines of code in this block will never be executed.

```suggestion
		$this->query(
			'',
			'SET NAMES {string:db_character_set}',
			[
				'db_character_set' => $this->character_set,
			],
		);
```

Reply:
As mentioned, if the connection fails it would still try to execute this.  I may have the if logic wrong if it doesn't make sense.

Reply:
`if (!empty($this->connection)) { ... }` would make a lot more sense, yes. In fact, I suggest putting everything from the `$this->get_version();` through to the end of the `SET NAMES` query inside an `if` block with that test, like so:

```
		if (!empty($this->connection)) {
			$this->get_version();
			$this->supports_pcre = version_compare($this->version, strpos($this->version, 'MariaDB') !== false ? '10.0.5' : '8.0.4', '>=');

			// For backward compatibility.
			if (!is_object(self::$db_connection)) {
				self::$db_connection = $this->connection;
			}

			// Ensure database has UTF-8 as its default input charset.
			$this->query(
				'',
				'SET NAMES {string:db_character_set}',
				[
					'db_character_set' => $this->character_set,
				],
			);
		}
```

Reply:
Should do the same in `SMF\Db\APIs\PostgreSQL::__construct()`, too.

Reply:
Would it be better to just do something like:

```
if (empty($this->connection) {
    return;
}
```

No need to nest the rest, just bail at this point, no need to execute anything.  I believe my code was attempting to mimic the earlier code, but I didn't make the right changes in regards to handling the `$non_fatal`.  I think its best to ignore it and yes if we don't have a connection, just stop.

Maybe even this:

```
		if (empty($this->connection) && !$non_fatal) {
			ErrorHandler::displayDbError();
		}
/* ADDED THIS */
		elseif (empty($this->connection)) {
			return;
		}
```

Essentially make it so we do the following checks
1. Check if we are missing a connection and have a fatal connection, if both are true; display error
2. Check if we have a missing connection, if true; return, no further logic can be done.

Only other logic being skipped is the SSI logic, but this won't mater really since there is no database connection.

Reply:
Makes sense. See my new review.